### DISCUSSION TITLE
How to add patched copies of ROMs to an existing collection?

### DISCUSSION THREAD

**Discussion by fooness** [emmercm/igir#648] [Q&A]
I’m using `igir` very similar to: https://igir.io/usage/personal/

Unfortunately, though, I’m struggling with adding patched copies of ROMs to my existing collection.

```
  Create patched copies of ROMs in an existing collection, not overwriting existing files:
    igir copy extract --input ROMs/ --patch Patches/ --output ROMs/
```

via: https://igir.io/#how-do-i-run-igir

I thought I’d simply put the patches in the specified folder, and then the whole `igir copy extract test clean` process is run again; but even though the output of `igir` states that _“X patched games were written”_, the to-be-patched ROMs on the SD card seem to be unaffected; booting up one of the should-have-been-patched ROMs does not show any changes the patches should have introduced.

Here’s my exact command:

```
  npx igir@latest copy extract test clean \
    --dat "${src}/No-Intro*.zip" \
    --input "${src}/No-Intro/" \
    --patch "${src}/Patches/" \
    --output "${dir}/games/{mister}/" \
    --dir-letter \
    --clean-exclude "${dir}/games/*/*.*" \
    --clean-exclude "${dir}/games/*/Palettes/**" \
    --no-bios \
    --no-device \
    --no-unlicensed \
    --only-retail \
    --no-debug \
    --no-demo \
    --no-beta \
    --no-sample \
    --no-prototype \
    --no-test-roms \
    --no-aftermarket \
    --no-homebrew \
    --no-unverified \
    --no-bad \
    --single \
    --prefer-language EN \
    --prefer-region USA,WORLD,EUR,JPN \
    --prefer-revision-newer \
    --prefer-retail
```

When specifying `--overwrite`, then everythings works as expected, but the whole process is taking very(!) long, as supposedly all the ROMs are overwritten, i.e. _re-written_, even if there’s no changes between destination and target ROM.

**Comment by emmercm** [comment]
`--overwrite-invalid` can help you save some HDD writes, but I feel it's not what you want.

---

Is there an un-patched ROM in your output directory that you expect to be overwritten by a patched ROM with your command? Or is your output directory empty?

Picking one of the should-have-been-patched ROMs as an example, can you tell me:

- The input file's location, size, and checksum
- The patch file's location
- The output file's (the one that doesn't appear to be patched) location, size, and checksum

**Reply by fooness** [reply]
> […] `--overwrite-invalid` can help you save some HDD writes, but I feel it's not what you want.

Yeah … thinking some more about which behavior I’m looking for, it’s basically a “patch” command (in addition to `copy extract test clean`) or “--patch-only” (or similar) input option, which processes the `--input`, `--patch`, and `--output` input option but only updates (i.e. overwrites) ROMs which have a patch with their respective CRC32 sum in the provided location. Also the content of the provided `--patch` directory can change, e.g. in case you’d want to apply a different patch to the (original, not previously patched) ROM.

What do you think about that?

---

> Is there an un-patched ROM in your output directory that you expect to be overwritten by a patched ROM with your command?

Yes, this is the case until using either `--overwrite` or `--overwrite-invalid` with their _“side-effects”_ of, well, overwriting ROMs that do not need any _updating_, so to say.

> Picking one of the should-have-been-patched ROMs as an example, can you tell me:
> 
>     The input file's location, size, and checksum
>     The patch file's location
>     The output file's (the one that doesn't appear to be patched) location, size, and checksum

_NOTE: Random example for ROM and patch; the problem is not only with this ROM and patch._

### ROM
*“The input file's location, size, and checksum”*
```
# file size of the `.zip` archive (shouldn’t be important)
$ stat -c %s /Volumes/ARCHIVE/No-Intro/Nintendo\ -\ Super\ Nintendo\ Entertainment\ System/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).zip
702836

# checksum of the `.zip` archive (shouldn’t be important)
$ crc32 /Volumes/ARCHIVE/No-Intro/Nintendo\ -\ Super\ Nintendo\ Entertainment\ System/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).zip
b4c519d3

# unzip archives
[…]

# file size of the `.sfc` ROM
$ stat -c %s /Volumes/ARCHIVE/No-Intro/Nintendo\ -\ Super\ Nintendo\ Entertainment\ System/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
1048576

# checksum of the `.sfc` ROM
$ crc32 /Volumes/ARCHIVE/No-Intro/Nintendo\ -\ Super\ Nintendo\ Entertainment\ System/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
777aac2f
```

### Patch
*“The patch file's location”*
```
$ stat -c %s /Volumes/ARCHIVE/Patches/snes/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\)\ \[Redux\ and\ Retranslation\ v10.2.2\].zip
231585

# unzip archive
[…]

$ stat -c %s /Volumes/ARCHIVE/Patches/snes/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\)\ \[Redux\ and\ Retranslation\ v10.2.2\]/Legend\ of\ Zelda\,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\)\ 777aac2f.ips
298280
```

### `igir` initial run *without* `--patch` input option
```
$ stat -c %s /Volumes/MiSTerFPGA/games/SNES/L/Legend\ of\ Zelda,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
1048576

$ crc32 /Volumes/MiSTerFPGA/games/SNES/L/Legend\ of\ Zelda,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
777aac2f
```

### `igir` second run *with* `--patch` input option
*“The output file's (the one that doesn't appear to be patched) location, size, and checksum”*
```
$ stat -c %s /Volumes/MiSTerFPGA/games/SNES/L/Legend\ of\ Zelda,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
1048576

$ crc32 /Volumes/MiSTerFPGA/games/SNES/L/Legend\ of\ Zelda,\ The\ -\ A\ Link\ to\ the\ Past\ \(USA\).sfc
777aac2f
```

**Reply by emmercm** [reply]
Is the resulting file still not patched even with one of the `--overwrite` flags?

I wonder if it's because your patch file is resulting in two ROM candidates that both want to write to `Legend of Zelda, The - A Link to the Past (USA).sfc`. If you rename your patch file, does the issue still persist?

**Comment by emmercm** [comment]
Breaking out part of what you said @fooness into a separate thread, are you looking for a `--patch-only` command that would only write patched ROMs, and none that match the DAT? You could first run a regular `igir copy extract` _without_ overwriting, and then a subsequent `igir copy extract --patch <path> --overwrite`.

**Reply by fooness** [reply]
@emmercm Thank you, that’s an interesting idea — I’ll give that a try!
